packages/contract-processors/src/moonbeam.ts

```diff
@@ -0,0 +1,150 @@
+// Copyright 2020-2021 OnFinality Limited authors & contributors
+// SPDX-License-Identifier: Apache-2.0
+
+import {Log} from '@ethersproject/abstract-provider';
+import {isHexString, hexStripZeros} from '@ethersproject/bytes';
+import {ApiPromise} from '@polkadot/api';
+import {CodecHash} from '@polkadot/types/interfaces';
+import {
+  SubqlDatasourceProcessor,
+  SubqlCustomDatasource,
+  SubqlHandlerKind,
+  SubqlNetworkFilter,
+  SubstrateEvent,
+  SubqlEventFilter,
+  SecondLayerHandlerProcessor,
+} from '@subql/types';
+import {plainToClass} from 'class-transformer';
+import {
+  IsOptional,
+  IsString,
+  validateSync,
+  ValidatorConstraint,
+  ValidatorConstraintInterface,
+  Validate,
+} from 'class-validator';
+
+type TopicFilter = string | string[] | null | undefined;
+
+export type MoonbeamDatasource = SubqlCustomDatasource<'substrate/Moonbeam'>;
+
+export interface MoonbeamEventFilter {
+  address?: string;
+  topics?: [TopicFilter, TopicFilter, TopicFilter, TopicFilter];
+}
+
+export type MoonbeamEvent = Log;
+
+@ValidatorConstraint({name: 'topifFilterValidator', async: false})
+export class TopicFilterValidator implements ValidatorConstraintInterface {
+  validate(value: TopicFilter): boolean {
+    return (
+      !value ||
+      (typeof value === 'string'
+        ? isHexString(value)
+        : Array.isArray(value)
+        ? !value.find((v) => !isHexString(v))
+        : false)
+    );
+  }
+
+  defaultMessage(): string {
+    return 'Value must be either null, undefined, hex string or hex string[]';
+  }
+}
+
+class MoonbeamEventFilterImpl implements MoonbeamEventFilter {
+  @IsOptional()
+  @IsString()
+  address?: string;
+  @IsOptional()
+  @Validate(TopicFilterValidator, {each: true})
+  topics?: [TopicFilter, TopicFilter, TopicFilter, TopicFilter];
+}
+
+type RawEvent = {
+  address: string;
+  topics: string[];
+  data: string;
+};
+
+function polkadotHashToEthHash(original: CodecHash): string {
+  // TODO something like this https://github.com/PureStake/moonbeam/blob/master/client/rpc/txpool/src/lib.rs#L87
+  return original.toHex();
+}
+
+function hexStringEq(a: string, b: string): boolean {
+  if (!isHexString(a) || !isHexString(b)) {
+    throw new Error('Inputs are not hex strings');
+  }
+  return hexStripZeros(a) === hexStripZeros(b);
+}
+
+const EventProcessor: SecondLayerHandlerProcessor<SubqlHandlerKind.Event, MoonbeamEventFilter, MoonbeamEvent> = {
+  baseFilter: [{module: 'evm', method: 'Log'}],
+  baseHandlerKind: 'substrate/EventHandler' as any /*SubqlHandlerKind.Event*/,
+  transformer(original: SubstrateEvent, ds: MoonbeamDatasource): Log {
+    const [eventData] = original.event.data;
+
+    const log = {
+      ...(eventData.toJSON() as unknown as RawEvent),
```

Comment:
Might not be the best way, though I'm also not sure user's expectation. for example for a uint256, is it string or BN users are expected to see here

Reply:
If you know what the type is, then just forcefully convert into it.